File: 124. Binary Tree Maximum Path Sum/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int _maxVale;
    public int maxPathSum(TreeNode root) {
        _maxVale = Integer.MIN_VALUE;
        nodeValue(root);
        return _maxVale;
    }
    
    private int nodeValue(TreeNode node) {
        if (node == null) {
            return 0;
        }
        
        int connectionSum = node.val;
        int leftSum = nodeValue(node.left);
        int rightSum = nodeValue(node.right);
        
        connectionSum += Math.max(leftSum, rightSum);
        int fullSum = node.val + leftSum + rightSum;
        
        _maxVale = Math.max(_maxVale, Math.max(fullSum, Math.max(connectionSum, node.val)));
        
        return Math.max(connectionSum, node.val);
    }
}

File: 124. Binary Tree Maximum Path Sum/Solution.go

import (
	"math"
)

// type TreeNode struct {
// 	Val int
// 	Left *TreeNode     
// 	Right *TreeNode
// }



func maxPathSum(root *TreeNode) int {
    maxVal := math.MinInt
	

	NodeValue(root , &maxVal)
	return maxVal
}

func NodeValue(node *TreeNode , maxVal *int )  int {  
	if node == nil {
		return 0
	}
	connectionSum ,leftSum , rightSum := node.Val , 0 ,0
	
	leftSum = NodeValue(node.Left , maxVal)
	rightSum = NodeValue(node.Right , maxVal)

	connectionSum += max(leftSum , rightSum)
	fullSum := node.Val + rightSum + leftSum

	*maxVal = max( *maxVal , fullSum , connectionSum , node.Val)
	
	return max(connectionSum , node.Val)
}

func max(nums ...int) int {
    if len(nums) == 0 {
        return 0 // or return an error, depending on your use case
    }
    
    maxNum := nums[0]
    for i := 1; i < len(nums); i++ {
        if nums[i] > maxNum {
            maxNum = nums[i]
        }
    }
    return maxNum
}